core/src/main/scala/com/mesosphere/usi/core/revive/SuppressReviveHandler.scala

```diff
@@ -0,0 +1,192 @@
+package com.mesosphere.usi.core.revive
+
+import akka.NotUsed
+import akka.stream.OverflowStrategy
+import akka.stream.scaladsl.{Flow, Source}
+import com.mesosphere.mesos.client.MesosCalls
+import com.mesosphere.usi.core.models.{PodId, PodSpecUpdatedEvent, RunningPodSpec, TerminalPodSpec}
+import com.mesosphere.usi.core.util.RateLimiterFlow
+import com.mesosphere.usi.metrics.{Counter, Metrics}
+import com.typesafe.scalalogging.StrictLogging
+import org.apache.mesos.v1.scheduler.Protos.Call
+import org.apache.mesos.v1.{Protos => Mesos}
+
+import scala.collection.JavaConverters._
+import scala.concurrent.duration.FiniteDuration
+
+class SuppressReviveHandler(
+    initialFrameworkInfo: Mesos.FrameworkInfo,
+    frameworkId: Mesos.FrameworkID,
+    metrics: Metrics,
+    mesosCallFactory: MesosCalls,
+    debounceReviveInterval: FiniteDuration)
+    extends StrictLogging {
+
+  import SuppressReviveHandler._
+
+  require(defaultRoles.nonEmpty, "initialFramework rolls must be non-empty!")
+
+  private[this] val reviveCountMetric: Counter = metrics.counter("usi.mesos.calls.revive")
+  private[this] val suppressCountMetric: Counter = metrics.counter("usi.mesos.calls.suppress")
+
+  private def defaultRoles = initialFrameworkInfo.getRolesList.asScala
+
+  /**
+    * Given a stream of PodSpecUpdatedEvent (which gives us the signal of which pods want offers), we emit a snapshot
+    * describing all roles to which the framework is subscribed, and the set of podIds which want offers for each role.
+    */
+  private[revive] def reviveStateFromPodSpecs: Flow[PodSpecUpdatedEvent, PodIdsWantingRoles, NotUsed] =
```

Comment:
I have to admit that I'm not a big fan to split this logic out into a flow. The original `suppressAndRevive` was a simple easy to test method. What is the benefit of mixing the logic with streams?

Reply:
Debouncing